File: Solutions/Python/kth_largest_element_in_an_array.py
"""
    [MEDIUM]
    215. Kth Largest Element in an Array

    Concepts:
    - array
    - heap/priority queue

    Stats:
        Runtime | 101 ms    [Beats 27.28%]
        Memory  | 26.60 MB  [Beats 99.01%]
"""

import heapq

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        n = len(nums)
        nums_left = max(0, min(n - k, n))
        heapq.heapify(nums)
        
        while nums and nums_left:
            heapq.heappop(nums)
            nums_left -= 1

        return heapq.heappop(nums)
        

File: Solutions/Cpp/kth_largest_element_in_an_array.cpp
/*
    [MEDIUM]
    215. Kth Largest Element in an Array

    Concepts:
    - array
    - heap/priority queue

    Stats:
        Runtime | 28 ms     [Beats 64.99%]
        Memory  | 59.27 MB  [Beats 83.03%]
*/

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int n = nums.size();
        std::make_heap(nums.begin(), nums.end());

        while (!nums.empty() && --k > 0) {
            std::pop_heap(nums.begin(), nums.end());
            if (k > 0) nums.pop_back();
        }

        std::pop_heap(nums.begin(), nums.end());
        return nums.back();
    }
};